test/framework/src/main/java/org/elasticsearch/entitlement/bootstrap/TestEntitlementsRule.java

```diff
@@ -0,0 +1,358 @@
+                public void evaluate() throws Throwable {
+                    if (active.compareAndSet(false, true)) {
+                        try {
+                            POLICY_MANAGER.setActive(false == withoutEntitlements);
+                            POLICY_MANAGER.setTriviallyAllowingTestCode(false == withEntitlementsOnTestCode);
+                            if (entitledPackages != null) {
+                                assert entitledPackages.value().length > 0 : "No test packages specified in @EntitledTestPackages";
+                                POLICY_MANAGER.setEntitledTestPackages(entitledPackages.value());
+                            } else {
+                                POLICY_MANAGER.setEntitledTestPackages();
+                            }
+                            BASE_DIR_PATHS.keySet().retainAll(List.of(TEMP));
+                            POLICY_MANAGER.clearModuleEntitlementsCache();
+                            // evaluate the suite
+                            base.evaluate();
+                        } finally {
+                            POLICY_MANAGER.setActive(false);
+                            POLICY_MANAGER.setTriviallyAllowingTestCode(true);
+                            POLICY_MANAGER.setEntitledTestPackages();
+                            BASE_DIR_PATHS.keySet().retainAll(List.of(TEMP));
+                            POLICY_MANAGER.clearModuleEntitlementsCache();
+                            active.set(false);
+                        }
+                    } else {
+                        throw new AssertionError("TestPolicyManager doesn't support test isolation, test suits cannot be run in parallel");
+                    }
+                }
+            };
+        } else if (withEntitlementsOnTestCode) {
+            throw new AssertionError(
+                "Cannot use @WithEntitlementsOnTestCode on tests that are not configured to use entitlements for testing"
+            );
+        } else {
+            return base;
+        }
+    }
+
+    /**
+     * Creates a entitlement grant for node specific paths.
+     */
+    public Closeable newNodeGrant(Settings settings, Path configPath) {
+        if (POLICY_MANAGER == null) {
+            return () -> {}; // noop if not running with entitlements
+        }
+
+        var unwrappedConfigPath = configPath;
+        while (unwrappedConfigPath instanceof FilterPath fPath) {
+            unwrappedConfigPath = fPath.getDelegate();
+        }
+        NodeGrant nodeGrant = new NodeGrant(settings, unwrappedConfigPath, this::revokeGrant);
+        addGrant(nodeGrant);
+        return nodeGrant;
+    }
+
+    /**
+     * Revoke all open node grants.
+     */
+    public void revokeNodeGrants() {
+        BASE_DIR_PATHS.keySet().retainAll(List.of(TEMP));
+        POLICY_MANAGER.clearModuleEntitlementsCache();
+    }
+
+    private record NodeGrant(Settings settings, Path configPath, Consumer<NodeGrant> onClose) implements Closeable {
```

Comment:
nit: we haven't used the "grant" terminology anywhere else in entitlements, I think it adds confusion since it comes from security manager. can we call this something else, perhaps TestNodeEntitlementPaths?